Set ILLink defaults

sbomer opened the pull request:
These defaults root the non-framework assemblies.
@nguerrera @fadimounir @swaroop-sridhar PTAL

sbomer committed:
Set ILLink defaults

These defaults root the non-framework assemblies.

nguerrera commented on src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs:
```diff
@@ -311,6 +353,26 @@ private void RemoveRootDescriptor(XDocument project)
+
+            var target = new XElement(ns + "Target",
+                                      new XAttribute("AfterTargets", "_SetILLinkDefaults"),
+                                      new XAttribute("Name", "_EnableNonFrameworkTrimming"));
```
Why does the process for testing this involve a custom target running after a "private" sdk target. If this is a valid user scenario under test, then the project file needs to be expressible without this.

sbomer replied:
It's not testing a user scenario. The user scenario we want to enable is the one that roots non-fx assemblies. Some of the other tests are checking that the linker is actually running by testing its ability to strip some specific code from our test project - which doesn't happen in the default mode, but is how we tell that the linker is running end-to-end.

nguerrera replied:
So the non default modes are unsupported? I still don't get why it's not possible to switch modes using "public" properties and no targets.

sbomer replied:
Pretty much. Adding custom root assemblies or root descriptors will be supported, but we won't have a way to say "aggressive" and have that translate into a different set of specific illink options (at least in this preview). Someone who wants to do that will have to use the ILLink task directly or something similar.

nguerrera replied:
Can you add a comment to the test code here explaining why we're doing this?

sbomer replied:
Sure!

fadimounir commented:
LGTM.

One unrelated nit comment I had: can we rename `_ILLink` to just `ILLink'? It will make this piece of code look more consistent:
``` xml
  <Target Name="ComputeAndCopyFilesToPublishDirectory"
          DependsOnTargets="ComputeFilesToPublish;
                            _ILLink;
                            CreateReadyToRunImages;
                            CopyFilesToPublishDirectory" />
```

sbomer committed:
Rename _ILLink -> ILLink

sbomer committed:
Add comment about testing with non-framework trimming

sbomer committed:
Prevent linking portable apps and warn about linker usage

sbomer committed:
Ignore case in stdout comparison

sbomer commented:
I added the warning about usage. It points to https://aka.ms/illink, which I set up to go to https://github.com/mono/linker for the moment. Once we have a better landing page we can change this.

I also added an error that disables linking non-self-contained apps.

sbomer commented:
@nguerrera PTAnotherL.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:
```diff
@@ -22,18 +22,20 @@ Copyright (c) .NET Foundation. All rights reserved.
                       '$(_TargetFrameworkVersionWithoutV)' >= '3.0' And '$(TargetFrameworkIdentifier)' == '.NETCoreApp' "
           DependsOnTargets="_RunILLink">
 
+    <NETSdkWarning ResourceName="ILLinkWarning" />
```
I wonder if this should be information. It doesn't seems like a good experience that if you add PublishTrimmed to a project where you use msbuild's warnaserror, then you immediately get a failed build.

sbomer replied:
Thinking about how to do this... is it fine to use `<Message Text="..." Importance="High" />`, or do I need to worry about localization? Would you like me to introduce a `NETSdkInformation`?

sbomer replied:
I added `NETSdkInformation`.

swaroop-sridhar commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:
```diff
@@ -84,17 +86,41 @@ Copyright (c) .NET Foundation. All rights reserved.
+   <Target Name="_SetILLinkDefaults"
+           DependsOnTargets="_ComputeManagedAssembliesToLink">
+
+    <NETSdkError Condition="'$(SelfContained)' != 'true'" ResourceName="ILLinkNotSupportedError" />
```
Should we always cause an error on fw-dependent publishing? 

I understand that with the current defaults, there's no point in running the linker on fw-dependent builds. But it isn't an error.  Is there a way to override the default settings (as done in the tests) and run the linker on fw-dependent apps? 

Can you please add a comment clarifying why we disallow fw-dependent configuration?

sbomer replied:
The currently-supported ways to override defaults are:
- adding roots via `TrimmerRootDescriptor` and `TrimmerRootAssembly`, which doesn't make it any more useful for portable apps, and
- using ILLink task directly, explicitly specifying the set of roots.

Somebody could do what the tests do and remove items from the defaults, but it requires depending on a private target.

In the future we may consider supporting something like `IsTrimmable` on `PackageReference`s, but until then I think we should keep the error - I can't think of a reasonable condition to put it under with the current override mechanisms. And someone who is willing to use the task directly will likely be able to work around the error as well.

I'll add a comment indicating as much.

sbomer replied:
I also wanted to note more generally about overriding defaults - the out of band package had more high-level msbuild properties that mapped to different combinations of roots and linker actions. However it ended up with a bunch of logic in msbuild that wasn't tested and kept breaking. If we do introduce a high-level "mode" concept, we'll need to ensure test coverage of different modes, and probably make this concept known to illink itself.

swaroop-sridhar replied:
Thanks for the explanation @sbomer

sbomer committed:
Use NETSdkInformation for ILLink informational text

sbomer committed:
Add comment about portable apps

sbomer commented:
@nguerrera if you see this could you please merge? I'd like to get a core-sdk build for some manual end-to-end validation.

sbomer commented:
@nguerrera could you please merge this?
